File: tcc_backend/src/services/metalPrice.service.ts
import axios from 'axios';
import db from '../database';
import logger from '../utils/logger';

interface MetalPriceResponse {
  success: boolean;
  base: string;
  timestamp: number;
  rates: {
    [key: string]: number;
  };
}

interface CachedMetalPrice {
  id: string;
  metal_symbol: string;
  base_currency: string;
  price_per_ounce: string;
  price_per_gram: string;
  price_per_kilogram: string;
  api_timestamp: string;
  cached_at: Date;
  expires_at: Date;
  is_expired: boolean;
}

export class MetalPriceService {
  private static readonly API_KEY = process.env.METAL_PRICE_API_KEY || '';
  private static readonly API_URL = process.env.METAL_PRICE_API_URL || 'https://api.metalpriceapi.com/v1';
  private static readonly CACHE_TTL = parseInt(process.env.METAL_PRICE_CACHE_TTL || '86400'); // Default 24 hours in seconds
  private static readonly TROY_OUNCE_IN_GRAMS = 31.1034768;

  /**
   * Get cached metal price from database
   */
  private static async getCachedPrice(
    metal: string,
    baseCurrency: string
  ): Promise<CachedMetalPrice | null> {
    try {
      const result = await db.query<CachedMetalPrice>(
        'SELECT * FROM get_cached_metal_price($1, $2)',
        [metal, baseCurrency]
      );

      if (result.length === 0) {
        return null;
      }

      return result[0];
    } catch (error: any) {
      logger.error(`Error getting cached metal price: ${error.message}`);
      return null;
    }
  }

  /**
   * Save metal price to database cache
   */
  private static async savePriceToCache(
    metal: string,
    baseCurrency: string,
    pricePerOunce: number,
    timestamp: number
  ): Promise<void> {
    try {
      const pricePerGram = pricePerOunce / this.TROY_OUNCE_IN_GRAMS;
      const pricePerKilogram = pricePerOunce * (1000 / this.TROY_OUNCE_IN_GRAMS);

      await db.query(
        'SELECT upsert_metal_price($1, $2, $3, $4, $5, $6, $7)',
        [
          metal,
          baseCurrency,
          pricePerOunce,
          pricePerGram,
          pricePerKilogram,
          timestamp,
          this.CACHE_TTL,
        ]
      );

      logger.info(`Metal price cached: ${metal} in ${baseCurrency}`);
    } catch (error: any) {
      logger.error(`Error saving metal price to cache: ${error.message}`);
      // Don't throw error here - continue even if caching fails
    }
  }

  /**
   * Get live metal prices from API (with database caching)
   * @param metals - Array of metal symbols (e.g., ['XAU', 'XAG', 'XPT'] for Gold, Silver, Platinum)
   * @param baseCurrency - Base currency (default: 'SLL' for Sierra Leone Leone)
   */
  static async getLiveMetalPrices(
    metals: string[] = ['XAU', 'XAG', 'XPT'],
    baseCurrency: string = 'SLL'
  ): Promise<MetalPriceResponse> {
    try {
      // Check database cache for all requested metals
      const cachedPrices = await Promise.all(
        metals.map((metal) => this.getCachedPrice(metal, baseCurrency))
      );

      // Check if all metals have valid (non-expired) cache entries
      const allCached = cachedPrices.every((cached) => cached && !cached.is_expired);

      if (allCached) {
        logger.info('Returning cached metal prices from database');

        // Build response from cached data
        const rates: { [key: string]: number } = {};
        let timestamp = 0;

        cachedPrices.forEach((cached) => {
          if (cached) {
            rates[cached.metal_symbol] = parseFloat(cached.price_per_ounce);
            timestamp = Math.max(timestamp, parseInt(cached.api_timestamp));
          }
        });

        return {
          success: true,
          base: baseCurrency,
          timestamp,
          rates,
        };
      }

      // Fetch fresh data from API
      const symbols = metals.join(',');
      const url = `${this.API_URL}/latest`;

      logger.info(`Fetching metal prices from API: ${url} (${symbols})`);

      const response = await axios.get<MetalPriceResponse>(url, {
        params: {
          api_key: this.API_KEY,
          base: baseCurrency,
          currencies: symbols,
        },
        timeout: 10000, // 10 second timeout
      });

      if (!response.data.success) {
        throw new Error('Metal Price API returned unsuccessful response');
      }

      // Save each metal price to database cache
      await Promise.all(
        Object.entries(response.data.rates).map(([metal, price]) =>
          this.savePriceToCache(metal, baseCurrency, price, response.data.timestamp)
        )
      );

      logger.info('Metal prices fetched successfully and cached to database');
      return response.data;
    } catch (error: any) {
      logger.error(`Error fetching metal prices: ${error.message}`);

      // Try to return cached data, even if expired
      const cachedPrices = await Promise.all(
        metals.map((metal) => this.getCachedPrice(metal, baseCurrency))
      );

      const anyCached = cachedPrices.some((cached) => cached !== null);

      if (anyCached) {
        logger.warn('Returning expired cached metal prices due to API error');

        const rates: { [key: string]: number } = {};
        let timestamp = 0;

        cachedPrices.forEach((cached) => {
          if (cached) {
            rates[cached.metal_symbol] = parseFloat(cached.price_per_ounce);
            timestamp = Math.max(timestamp, parseInt(cached.api_timestamp));
          }
        });

        return {
          success: true,
          base: baseCurrency,
          timestamp,
          rates,
        };
      }

      throw new Error(`Failed to fetch metal prices: ${error.message}`);
    }
  }

  /**
   * Get price for a specific metal
   * @param metal - Metal symbol (e.g., 'XAU' for Gold)
   * @param baseCurrency - Base currency
   */
  static async getMetalPrice(
    metal: string,
    baseCurrency: string = 'SLL'
  ): Promise<{ metal: string; price: number; currency: string; timestamp: number }> {
    try {
      const prices = await this.getLiveMetalPrices([metal], baseCurrency);

      if (!prices.rates[metal]) {
        throw new Error(`Price not found for metal: ${metal}`);
      }

      return {
        metal,
        price: prices.rates[metal],
        currency: baseCurrency,
        timestamp: prices.timestamp,
      };
    } catch (error: any) {
      logger.error(`Error fetching price for ${metal}: ${error.message}`);
      throw error;
    }
  }

  /**
   * Convert metal price to different units
   * Metal prices are typically per troy ounce (31.1g)
   * @param pricePerOunce - Price per troy ounce
   * @param unit - Target unit ('gram', 'ounce', 'kilogram')
   */
  static convertMetalPriceUnit(
    pricePerOunce: number,
    unit: 'gram' | 'ounce' | 'kilogram'
  ): number {
    switch (unit) {
      case 'gram':
        return pricePerOunce / this.TROY_OUNCE_IN_GRAMS;
      case 'ounce':
        return pricePerOunce;
      case 'kilogram':
        return pricePerOunce * (1000 / this.TROY_OUNCE_IN_GRAMS);
      default:
        return pricePerOunce;
    }
  }

  /**
   * Clear database cache (useful for testing or forced refresh)
   */
  static async clearCache(): Promise<void> {
    try {
      await db.query('DELETE FROM metal_price_cache');
      logger.info('Metal price database cache cleared');
    } catch (error: any) {
      logger.error(`Error clearing metal price cache: ${error.message}`);
      throw error;
    }
  }

  /**
   * Get formatted metal prices with percentage changes
   * Note: Percentage change calculation would require historical data
   */
  static async getFormattedMetalPrices(baseCurrency: string = 'SLL'): Promise<{
    gold: { price: number; pricePerGram: number };
    silver: { price: number; pricePerGram: number };
    platinum: { price: number; pricePerGram: number };
    timestamp: number;
  }> {
    try {
      const prices = await this.getLiveMetalPrices(['XAU', 'XAG', 'XPT'], baseCurrency);

      return {
        gold: {
          price: prices.rates.XAU || 0,
          pricePerGram: this.convertMetalPriceUnit(prices.rates.XAU || 0, 'gram'),
        },
        silver: {
          price: prices.rates.XAG || 0,
          pricePerGram: this.convertMetalPriceUnit(prices.rates.XAG || 0, 'gram'),
        },
        platinum: {
          price: prices.rates.XPT || 0,
          pricePerGram: this.convertMetalPriceUnit(prices.rates.XPT || 0, 'gram'),
        },
        timestamp: prices.timestamp,
      };
    } catch (error: any) {
      logger.error(`Error getting formatted metal prices: ${error.message}`);
      throw error;
    }
  }
}

File: tcc_user_mobile_client/lib/services/metal_price_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/metal_price_model.dart';
import '../config/app_constants.dart';

class MetalPriceService {
  // Conversion constants
  static const double gramsPerTroyOunce = 31.1035;

  /// Get live metal prices (Gold, Silver, Platinum) from CurrencyBeacon
  /// Returns formatted prices in the specified base currency
  Future<Map<String, dynamic>> getMetalPrices({
    String baseCurrency = 'USD',
  }) async {
    try {
      // Fetch metal prices from CurrencyBeacon
      // Metals: XAU (Gold), XAG (Silver), XPT (Platinum)
      String url = '${AppConstants.currencyBeaconBaseUrl}/latest?api_key=${AppConstants.currencyBeaconApiKey}&base=$baseCurrency&symbols=XAU,XAG,XPT';

      final response = await http.get(
        Uri.parse(url),
        headers: {'Accept': 'application/json'},
      ).timeout(Duration(seconds: AppConstants.apiTimeout));

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);

        if (data['response'] != null && data['response']['rates'] != null) {
          final rates = data['response']['rates'] as Map<String, dynamic>;

          // CurrencyBeacon returns rates as "1 USD = X troy oz"
          // We need to invert to get "price per troy oz in USD"
          final xauRate = (rates['XAU'] as num?)?.toDouble() ?? 0.0;
          final xagRate = (rates['XAG'] as num?)?.toDouble() ?? 0.0;
          final xptRate = (rates['XPT'] as num?)?.toDouble() ?? 0.0;

          // Calculate prices per troy ounce
          final goldPricePerOz = xauRate > 0 ? 1 / xauRate : 0.0;
          final silverPricePerOz = xagRate > 0 ? 1 / xagRate : 0.0;
          final platinumPricePerOz = xptRate > 0 ? 1 / xptRate : 0.0;

          // Calculate prices per gram (1 troy oz = 31.1035 grams)
          final goldPricePerGram = goldPricePerOz / gramsPerTroyOunce;
          final silverPricePerGram = silverPricePerOz / gramsPerTroyOunce;
          final platinumPricePerGram = platinumPricePerOz / gramsPerTroyOunce;

          final metalPricesResponse = MetalPricesResponse(
            base: baseCurrency.toUpperCase(),
            metals: MetalPrices(
              gold: MetalPriceDetail(
                price: goldPricePerOz,
                pricePerGram: goldPricePerGram,
              ),
              silver: MetalPriceDetail(
                price: silverPricePerOz,
                pricePerGram: silverPricePerGram,
              ),
              platinum: MetalPriceDetail(
                price: platinumPricePerOz,
                pricePerGram: platinumPricePerGram,
              ),
            ),
            timestamp: DateTime.now().millisecondsSinceEpoch ~/ 1000,
          );

          return {
            'success': true,
            'data': metalPricesResponse,
          };
        }
      }

      return {
        'success': false,
        'error': 'Failed to fetch metal prices from CurrencyBeacon',
      };
    } catch (e) {
      return {
        'success': false,
        'error': 'Metal prices error: ${e.toString()}',
      };
    }
  }

  /// Get specific metal price from CurrencyBeacon
  /// [metal] - Metal symbol (XAU, XAG, XPT)
  /// [baseCurrency] - Base currency code (default: USD)
  /// [unit] - Unit for price (gram, ounce, kilogram)
  Future<Map<String, dynamic>> getMetalPrice({
    required String metal,
    String baseCurrency = 'USD',
    String unit = 'ounce',
  }) async {
    try {
      String url = '${AppConstants.currencyBeaconBaseUrl}/latest?api_key=${AppConstants.currencyBeaconApiKey}&base=$baseCurrency&symbols=${metal.toUpperCase()}';

      final response = await http.get(
        Uri.parse(url),
        headers: {'Accept': 'application/json'},
      ).timeout(Duration(seconds: AppConstants.apiTimeout));

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);

        if (data['response'] != null && data['response']['rates'] != null) {
          final rates = data['response']['rates'] as Map<String, dynamic>;
          final metalRate = (rates[metal.toUpperCase()] as num?)?.toDouble() ?? 0.0;

          // Calculate price per troy ounce
          final pricePerOz = metalRate > 0 ? 1 / metalRate : 0.0;
          final pricePerGram = pricePerOz / gramsPerTroyOunce;

          // Calculate price based on unit
          double price = pricePerOz;
          if (unit == 'gram') {
            price = pricePerGram;
          } else if (unit == 'kilogram') {
            price = pricePerGram * 1000;
          }

          final metalPrice = MetalPrice(
            metal: metal.toUpperCase(),
            price: price,
            pricePerGram: pricePerGram,
            currency: baseCurrency.toUpperCase(),
            timestamp: DateTime.now().millisecondsSinceEpoch ~/ 1000,
          );

          return {
            'success': true,
            'data': metalPrice,
          };
        }
      }

      return {
        'success': false,
        'error': 'Failed to fetch metal price from CurrencyBeacon',
      };
    } catch (e) {
      return {
        'success': false,
        'error': 'Metal price error: ${e.toString()}',
      };
    }
  }

  /// Format metal prices for display on home screen
  /// Returns a list of formatted metal data ready for UI
  Future<List<Map<String, dynamic>>> getFormattedMetalPricesForDisplay({
    String baseCurrency = 'SLL',
  }) async {
    try {
      final result = await getMetalPrices(baseCurrency: baseCurrency);

      if (result['success'] != true) {
        return [];
      }

      final metalPricesResponse = result['data'] as MetalPricesResponse;
      final metals = metalPricesResponse.metals;

      return [
        {
          'name': 'Gold',
          'code': 'XAU',
          'price': metals.gold.pricePerGram,
          'pricePerOunce': metals.gold.price,
          'currency': baseCurrency,
        },
        {
          'name': 'Silver',
          'code': 'XAG',
          'price': metals.silver.pricePerGram,
          'pricePerOunce': metals.silver.price,
          'currency': baseCurrency,
        },
        {
          'name': 'Platinum',
          'code': 'XPT',
          'price': metals.platinum.pricePerGram,
          'pricePerOunce': metals.platinum.price,
          'currency': baseCurrency,
        },
      ];
    } catch (e) {
      return [];
    }
  }
}
